### DISCUSSION TITLE
rsync from mox to local machine?

### DISCUSSION THREAD

**Discussion by aspencoyle** [sr320/course-fish546-2021#94] [Q&A] [answered]
I successfully ran a slurm script (!!!) to perform blastn on a .fasta file! However, when I try to get the resulting files onto my local machine, it just isn't working!

When I run `rsync -a afcoyle@mox1:/gscratch/scrubbed/afcoyle/ \ /mnt/c/Users/acoyl/Documents/GitHub/hemat_bairdi_transcriptome/testdir`, it returns 
```
rsync: mkdir "/mnt/c/Users/acoyl/Documents/GitHub/hemat_bairdi_transcriptome/testdir" failed: No such file or directory (2)
rsync error: error in file IO (code 11) at main.c(657) [Receiver=3.1.2]
```

I'm working from WSL, so I assumed that might be the issue. However, when I replaced it with the Windows filepath, as follows:
```
rsync -a afcoyle@mox1:/gscratch/scrubbed/afcoyle \Users\acoyl\Documents\GitHub\hemat_bairdi_transcriptome\testdir
```

it didn't return an error, but it didn't copy anything over - the directory stayed empty!

Any ideas on what I'm screwing up here?

**Comment by kubu4** [comment]
A couple of things to try:

1. Address the error message: `rsync: mkdir "/mnt/c/Users/acoyl/Documents/GitHub/hemat_bairdi_transcriptome/testdir" failed: No such file or directory (2)`

    Make sure the `/mnt/c/Users/acoyl/Documents/GitHub/hemat_bairdi_transcriptome/testdir` directory exists before attempting to transfer.

1. Leave the trailing backslash off of your source path, like this:

`rsync -a afcoyle@mox1:/gscratch/scrubbed/afcoyle /mnt/c/Users/acoyl/Documents/GitHub/hemat_bairdi_transcriptome/testdir`

**Reply by aspencoyle** [reply]
Hmm, no luck with either one! I tried Steven's suggestion, but have a hard time logging into Gannett. I've persistently had problems with Gannet - it just doesn't seem to like me! For instance, my password (which is definitely correct and works just fine when accessing via the website) is rejected when I try to log in via the command line. 

Attempt to access Gannet: `ssh -v afcoyle@gannet.fish.washington.edu`

Part of the issue might be that I may have multiple accounts - when I log in online, it gives me multiple options - one with the username "afcoyle" and the other with "afcoyle@uw.edu". When I checked my password manager, the two have separate passwords.

Anyway, any clue why I'm having this trouble with rsync or these problems with Gannet?

**Reply by kubu4** [reply]
> Attempt to access Gannet: ssh -v afcoyle@gannet.fish.washington.edu
>
> Part of the issue might be that I may have multiple accounts - when I log in online, it gives me multiple options - one with the username "afcoyle" and the other with "afcoyle@uw.edu". When I checked my password manager, the two have separate passwords.
>
> Anyway, any clue why I'm having this trouble with rsync or these problems with Gannet?

Please start a separate issue for this.

**Reply by kubu4** [reply]
> Hmm, no luck with either one

Could you please elaborate (and, possibly show your commands/outputs/errors)?

1. Did you create your destination directory? If yes, please run `pwd` command in your destination and paste command/results in reply.

2. When you ran `rsync` after creating your desired destination directory did you get an error message? If yes, please paste the command/results in reply.

3. What command(s) did you use to transfer files to Mox?

**Reply by kubu4** [reply]
Also, please try to transfer just a single file from Mox to your desired destination. Paste command/results.

**Reply by aspencoyle** [reply]
Sounds good, I'll start a separate issue after replying!

> Did you create your destination directory? If yes, please run pwd command in your destination and paste command/results in reply.

Yep, the directory had already been created. `pwd` returns the following:
`/mnt/c/Users/acoyl/Documents/GitHub/hemat_bairdii_transcriptome/testdir`

>When you ran rsync after creating your desired destination directory did you get an error message? If yes, please paste the command/results in reply.

Sure thing - here's my command and results
```
[afcoyle@mox1 afcoyle]$ rsync -a afcoyle@mox1:/gscratch/scrubbed/afcoyle /mnt/c/Users/acoyl/Documents/GitHub/hemat_bairdii_transcriptome/testdir
Password:
Enter passcode or select one of the following options:

 1. Duo Push to Android (XXX-XXX-3428)
 2. Phone call to Android (XXX-XXX-3428)

Duo passcode or option [1-2]: 1
rsync: mkdir "/mnt/c/Users/acoyl/Documents/GitHub/hemat_bairdii_transcriptome/testdir" failed: No such file or directory (2)
rsync error: error in file IO (code 11) at main.c(657) [Receiver=3.1.2]
```

>What command(s) did you use to transfer files to Mox?

I downloaded the files from GitHub using `curl` (just two files and one wasn't on my local machine, figured that'd be easiest)

>Also, please try to transfer just a single file from Mox to your desired destination. Paste command/results.

Sure thing - here you go! I tried to transfer my slurm script over, no luck!
```
[afcoyle@mox1 afcoyle]$ rsync -a afcoyle@mox1:/gscratch/scrubbed/afcoyle/coyle_practice_slurmscript.txt /mnt/c/Users/aco
yl/Documents/GitHub/hemat_bairdii_transcriptome/testdir
Password:
Enter passcode or select one of the following options:

 1. Duo Push to Android (XXX-XXX-3428)
 2. Phone call to Android (XXX-XXX-3428)

Duo passcode or option [1-2]: 1
rsync: change_dir#3 "/mnt/c/Users/acoyl/Documents/GitHub/hemat_bairdii_transcriptome" failed: No such file or directory (2)
rsync error: errors selecting input/output files, dirs (code 3) at main.c(695) [Receiver=3.1.2]
```

**Reply by kubu4** [reply]
Hmmm, I tested out rsync from your Mox directory to my computer (using WSL) and rsync'd fine. 

Can you please check the permissions on your destination directory?

``` bash
cd /mnt/c/Users/acoyl/Documents/GitHub/hemat_bairdii_transcriptome
ls -l
```

**Comment by sr320** [comment, marked as answer]
Just an FYI - I move everything from mox to gannet first. Keeps an archive of everything..

eg 
```
sr320@Gannet:/volume2/web/seashell/bu-mox
rsync -avz --progress sr320@mox.hyak.uw.edu:/gscratch/scrubbed/sr320/ scrubbed/
```

https://gannet.fish.washington.edu/seashell/bu-mox/

**Comment by aspencoyle** [comment]
Alright, I got gannet figured out (thanks, Sam!) and tried transferring the whole folder from mox to gannet. Here's what I got when attempting to transfer the whole folder:

```
 afcoyle@Gannet:/volume2/web/seashell/bu-mox$ 
rsync -avz --progress afcoyle@mox.hyak.uw.edu:/gscratch/scrubbed/afcoyle scrubbed/

The authenticity of host 'mox.hyak.uw.edu (198.48.92.25)' can't be established.
ECDSA key fingerprint is SHA256:Jk4oBsB8yvl+/8ftxsajCfEa1p8FEKt6oRh2WpZ/Ozo.
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added 'mox.hyak.uw.edu,198.48.92.25' (ECDSA) to the list of known hosts.
Password:
Password:
Enter passcode or select one of the following options:

 1. Duo Push to Android (XXX-XXX-3428)
 2. Phone call to Android (XXX-XXX-3428)

Duo passcode or option [1-2]: 1
receiving incremental file list
rsync: recv_generator: mkdir "/volume2/web/seashell/bu-mox/scrubbed/afcoyle" failed: Permission denied (13)
*** Skipping any contents from this failed directory ***
afcoyle/

sent 42 bytes  received 449 bytes  17.23 bytes/sec
total size is 352749673  speedup is 718431.11
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1904) [generator=3.0.9]
```

When I checked the scrubbed folder, no files actually transferred over to Gannet! Tried again with just one file. Again, no luck!

```
afcoyle@Gannet:/volume2/web/seashell/bu-mox$ rsync -avz --progress afcoyle@mox.hyak.uw.edu:/gscratch/scrubbed/afcoyle/coyle_practice_slurmscript.txt scrubbed/
Password:
Enter passcode or select one of the following options:

 1. Duo Push to Android (XXX-XXX-3428)
 2. Phone call to Android (XXX-XXX-3428)

Duo passcode or option [1-2]: 1
receiving incremental file list
rsync: mkstemp "/volume2/web/seashell/bu-mox/scrubbed/.coyle_practice_slurmscript.txt.5JJZUK" failed: Permission denied (13)
         818 100%  798.83kB/s    0:00:00 (total: 100%) (xfer#1, to-check=0/1)

sent 42 bytes  received 565 bytes  39.16 bytes/sec
total size is 818  speedup is 1.35
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1904) [generator=3.0.9]
```

It seems to me that this might be a permissions issue! I figured it might be good to ask for help here if so - don't want to noodle around with permissions on Mox and accidentally screw things up. Here are the permissions for all files in my mox folder

```
[afcoyle@mox1 afcoyle]$ ls -l
total 0
-rw-r--r-- 1 afcoyle hyak-srlab  818 Feb 23 20:23 coyle_practice_slurmscript.txt
drwxr-xr-x 2 afcoyle hyak-srlab 4096 Feb 23 20:29 data
drwxr-xr-x 3 afcoyle hyak-srlab 4096 Feb 23 20:23 output
-rw-r--r-- 1 afcoyle hyak-srlab  153 Feb 23 20:30 slurm-1303821.out
```

**Reply by kubu4** [reply]
> `/volume2/web/seashell/bu-mox`

Permission denied because this is @sr320's folder. He was just providing an example above. You'll need to `rsync` to somewhere in your own folder on Gannet.

**Reply by aspencoyle** [reply]
Fantastic, got it to work! Created a new folder on Gannet (theme seemed to be "scientific name of species of choice", so I went with an homage to the noble sockeye salmon and named it nerka). Ran the following command on Gannet and it ran perfectly: `rsync -avz --progress afcoyle@mox.hyak.uw.edu:/gscratch/scrubbed/afcoyle scrubbed/`

Then transferred the files on to my local machine by running the following (from my local machine, not from Gannet)
```
rsync -chavzP --stats \
afcoyle@gannet.fish.washington.edu:/volume2/web/nerka/mox_transfers/scrubbed/2021_02_24_practiceslurm \
/mnt/c/Users/acoyl/Documents/GitHub/hemat_bairdii_transcriptome/testdir
```

And that worked perfectly too! 

Anyway, thanks a ton for both of you for the help!
